### PR TITLE
Increase laziness of LazyList.cons.apply

### PR DESCRIPTION
**Pull request by NthPortal**
Change `LazyList.cons.apply` so that the tail computation
is wrapped as a state computation for a new `LazyList`,
thus deferring its execution as long as possible.

Follow-up to #8985

--------

Before (Scala 2.13.3-):

```scala
scala 2.13.3> LazyList.cons(1, { println("hey"); LazyList() })
val res6: scala.collection.immutable.LazyList[Int] = LazyList(<not computed>)

scala 2.13.3> res6.head
hey
val res7: Int = 1
```

After (Scala 2.13.4+):

```scala
scala 2.13.4> LazyList.cons(1, { println("hey"); LazyList() })
val res6: scala.collection.immutable.LazyList[Int] = LazyList(<not computed>)

scala 2.13.4> res6.head
val res7: Int = 1
```

### CODE REVIEW

**Review by julienrf — APPROVED** [review]

File: test/junit/scala/collection/immutable/LazyListLazinessTest.scala [hunk 1]
```diff
@@ -693,6 +693,20 @@ class LazyListLazinessTest {
     assertLazyAllSkipping(op, 4)
   }
 
+  private def genericCons_unapply_properlyLazy(unapply: LazyList[Int] => Option[(Int, LazyList[Int])]): Unit = {
```
Comment (NthPortal) [inline test/junit/scala/collection/immutable/LazyListLazinessTest.scala:696]: I noticed there were no tests for these while I was in the area, so I added tests

**Review by SethTisue — APPROVED** [review]

**Comment by SethTisue** [comment]
@NthPortal would you mind improving the PR title and description to be at least a little more informative for end users? since we're linking to this from the 2.13.4 release notes

perhaps you can borrow some verbiage from the discussion on #8985

**Comment by NthPortal** [comment]
@SethTisue is that okay?

**Comment by SethTisue** [comment]
@NthPortal better — if you want to improve it further, I would suggest adding an example code snippet that is affected by the change, and describe how it is affected. even though I've looked at this stuff multiple times before, when I see `LazyList.cons.apply` my reaction is still "huh? what's that?" and I have to go look it up. an example would really help, I think.

**Comment by som-snytt** [comment]
"If you're as lazy as LazyList, you probably don't need to know more."

Sometimes it helps to be usefully vague.

Do I need to learn this?

- [ ] No.

**Pull request merged**